server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java

```diff
@@ -0,0 +1,463 @@
+            throw new ParsingException(parser.getTokenLocation(), "Malformed IntervalsSource definition, expected start_object after source name");
+        }
+        IntervalsSourceProvider provider = parser.namedObject(IntervalsSourceProvider.class, sourceType, null);
+        //end_object of the specific query (e.g. match, multi_match etc.) element
+        if (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(),
+                "[" + sourceType + "] malformed source, expected [END_OBJECT] but found [" + parser.currentToken() + "]");
+        }
+        //end_object of the query object
+        if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(),
+                "[" + sourceType + "] malformed source, expected [END_OBJECT] but found [" + parser.currentToken() + "]");
+        }
+        return provider;
+    }
+
+    public static final IntervalsSource NO_INTERVALS = new IntervalsSource() {
+
+        @Override
+        public IntervalIterator intervals(String field, LeafReaderContext ctx) throws IOException {
+            return new IntervalIterator() {
+                @Override
+                public int start() {
+                    return NO_MORE_INTERVALS;
+                }
+
+                @Override
+                public int end() {
+                    return NO_MORE_INTERVALS;
+                }
+
+                @Override
+                public int nextInterval() throws IOException {
+                    return NO_MORE_INTERVALS;
+                }
+
+                @Override
+                public float matchCost() {
+                    return 0;
+                }
+
+                @Override
+                public int docID() {
+                    return NO_MORE_DOCS;
+                }
+
+                @Override
+                public int nextDoc() throws IOException {
+                    return NO_MORE_DOCS;
+                }
+
+                @Override
+                public int advance(int target) throws IOException {
+                    return NO_MORE_DOCS;
+                }
+
+                @Override
+                public long cost() {
+                    return 0;
+                }
+            };
+        }
+
+        @Override
+        public void extractTerms(String field, Set<Term> terms) {
+
+        }
+
+        @Override
+        public int hashCode() {
+            return 0;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            return other == this;
+        }
+
+        @Override
+        public String toString() {
+            return "no_match";
+        }
+    };
+
+    public static class Match extends IntervalsSourceProvider {
+
+        public static final String NAME = "match";
+
+        private final String text;
+        private final int maxWidth;
```

Comment:
Should we extract this option to have a dedicated source for max_width ? Or could it be handled by a scripted intervals source ?